File: 2019/levelUpBatch_00/lecture_010_HashMapAndHeap/heapConstruction/priorityQueue.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> pq;
bool isMax = true;

void display()
{
    for (int ele : pq)
        cout << ele << " ";
    cout << endl;
}

bool compareTo(int a, int b)
{
    if (isMax)
        return a > b;
    else
        return a < b;
}

void upheapify(int ci)
{
    int pi = (ci - 1) / 2;
    if (pi >= 0 && compareTo(pq[ci], pq[pi]))
    {
        swap(pq[pi], pq[ci]);
        upheapify(pi);
    }
}

void downHeapify(int pi)
{
    int lci = 2 * pi + 1;
    int rci = 2 * pi + 2;
    int maxi = pi;

    if (lci < pq.size() && compareTo(pq[lci], pq[maxi]))
        maxi = lci;
    if (rci < pq.size() && compareTo(pq[rci], pq[maxi]))
        maxi = rci;

    if (maxi != pi)
    {
        swap(pq[maxi], pq[pi]);
        downHeapify(maxi);
    }
}

void push(int val)
{
    pq.push_back(val);
    upheapify(pq.size() - 1);
}

void pop()
{
    int n = pq.size() - 1;
    swap(pq[0], pq[n]);
    pq.erase(pq.begin() + n);
    downHeapify(0);
}

int top()
{
    return pq[0];
}

void SetPriorityQueue(vector<int> &arr)
{
    for (int ele : arr)
        pq.push_back(ele);

    for (int i = pq.size() - 1; i >= 0; i--)
        downHeapify(i);
}

void set1()
{
    vector<int> arr{10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13};
    isMax=false;
    SetPriorityQueue(arr);

    while (pq.size() != 0)
    {
        cout << top() << " ";
        // display();
        pop();
    }
}

void solve()
{
    set1();
}

int main()
{
    solve();
    return 0;
}

File: 2019/BPIT_PP_Batch/lecture_007_HashMapHeap/heap/priorityqueue.java
import java.util.ArrayList;
public class priorityqueue{

    private ArrayList<Integer> pq;
    private boolean isMaxHeap = true;

    // Constructor.===========================================

    public priorityqueue(int[] arr,boolean isMax){
        intialize(isMax);
        for(int ele : arr) pq.add(ele);
        constructHeap();
    }

    public priorityqueue(){
        intialize(true);   
    }

    private void intialize(boolean isMax){
        pq = new ArrayList<>();
        isMaxHeap = isMax;
    }

    private void constructHeap(){
        for(int i = pq.size() - 1; i>=0 ;i--)
           downHeapify(i);
    }

    private int compareTo(int i,int j){
        if(isMaxHeap) return pq.get(i) - pq.get(j);
        else return pq.get(j) - pq.get(i);
    }

    private void swap(int i,int j){
        int v1 = pq.get(i);
        int v2 = pq.get(j);

        pq.set(i,v2);
        pq.set(j,v1);
    } 

    private void downHeapify(int pi){  // O(log(n))
        int maxidx = pi;
        int lci = (pi << 1) + 1;
        int rci = (pi << 1) + 2;

        if(lci < pq.size() && compareTo(lci,maxidx) > 0 ) maxidx = lci;
        if(rci < pq.size() && compareTo(rci,maxidx) > 0) maxidx = rci;

        if(maxidx != pi){
            swap(maxidx,pi);
            downHeapify(maxidx);
        }
    }

    private void upHeapify(int ci){  // O(log(n))
        int pi = (ci - 1) >> 1;
        if(pi >=0 && compareTo(ci,pi) > 0){
            swap(ci,pi);
            upHeapify(pi);
        }
    }

    //=====================================================================

    public int size(){
        return pq.size();
    }

    public boolean isEmpty(){
        return pq.size() == 0;
    }

    public void add(int val){ // O(log(n))
        pq.add(val);
        upHeapify(pq.size()-1);
    }

    public int peek() throws Exception{ // O(1)
        if(pq.size() == 0){
            throw new Exception("NullPointerException");
        }

        return pq.get(0);
    }

    public int poll() throws Exception{  // O(log(n))
        if(pq.size() == 0){
            throw new Exception("NullPointerException");
        }
        
        for(int ele : pq) System.out.print(ele + " ");
        System.out.println();
        
        int rv = pq.get(0);
        swap(0, pq.size() - 1);

        pq.remove(pq.size() - 1);
        downHeapify(0);

        return rv;
    }

    // public void update(int idx){
    //     downHeapify(idx);
    //     upHeapify(idx);
    // }   
}